design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,142 @@
+# KLIP 52 - BYTES Data Type Support
+
+**Author**: Zara Lim (@jzaralim) | 
+**Release Target**: 0.21 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7764
+
+**tl;dr:** _Add support for the BYTES data type. This will allow users to work with BLOBs of data that don't fit into any other data type._
+           
+## Motivation and background
+
+Currently, ksqlDB can only handle a set of primitive types and combinations of them.
+A BYTES data type would allow users to work with data that does not fit into any of
+the primitive types such as images, as well as BLOB/binary data from other databases.
+
+## What is in scope
+* Add BYTES type to KSQL
+* Support BYTE comparisons
+* Support BYTES usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of BYTES to Avro, JSON, Protobuf and Delimited formats
+* Adding/updating UDFs to support the BYTES type
+
+## What is not in scope
+* Fixed sized BYTES (`BYTES(3)` representing 3 bytes, for example) - This is supported by Kafka Connect by adding the `connect.fixed.size`
+key in a bytes schema, but this will not be included in this KLIP.
+
+## Public APIS
+
+### BYTES
+
+The BYTES data type will store an array of raw bytes of an unspecified length. The maximum size of
+the array is limited by the maximum size of a Kafka message, as well as possibly by the value format being used.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (b BYTES, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field BYTES>) AS ...
+```
+
+By default, BYTES will be displayed in console as HEX strings, where each byte is represented by two characters.
```

Comment:
> By default, BYTES will be displayed in console as HEX strings

* Does "console" refer strictly to the ksqlDB CLI? If so, the example above should use the `ksql>` prompt.
* Are we choosing hex here for consistency with other databases?
* What about the REST API? With JSON, it's typical to send byte arrays as base64 strings.
* Should the Java client automatically convert their wire representation into ByteBuffer objects?

Reply:
👍 just saw later in the KLIP that JSON/delimited will use base64. That covers the REST question above.

Reply:
* Yeah it's the CLI. The two common CLI representations in other databases are hex and UTF-8. IMO hex is easier to use with because sometimes UTF-8 representations show weird characters.
* The Java client should keep base64 strings as is until a user wants it converted to ByteBuffer (if they call `getBytes` on the response object.)

Reply:
Added some clarifications to the KLIP on how BYTES will be handled by the REST API/Java client.